x/ecocredit/core/genesis.go

```diff
@@ -0,0 +1,267 @@
+	if err != nil {
+		return err
+	}
+	defer cItr.Close()
+
+	// make sure credit type exist for class abbreviation in params
+	for cItr.Next() {
+		class, err := cItr.Value()
+		if err != nil {
+			return err
+		}
+
+		if _, ok := abbrevToPrecision[class.CreditType]; !ok {
+			return sdkerrors.ErrNotFound.Wrapf("credit type not exist for %s abbreviation", class.CreditType)
+		}
+	}
+
+	classIds := make(map[uint64]uint64) // map of projectID to classID
+	pItr, err := ss.ProjectInfoTable().List(ormCtx, api.ProjectInfoPrimaryKey{})
+	if err != nil {
+		return err
+	}
+	defer pItr.Close()
+
+	for pItr.Next() {
+		val, err := pItr.Value()
+		if err != nil {
+			return err
+		}
+
+		if _, exists := classIds[val.Id]; exists {
+			continue
+		}
+		classIds[val.Id] = val.ClassId
+	}
+
+	decimalPlaces := make(map[uint64]uint32) // map of batchID to precision
+	bItr, err := ss.BatchInfoTable().List(ormCtx, api.BatchInfoPrimaryKey{})
+	if err != nil {
+		return err
+	}
+	defer bItr.Close()
+
+	for bItr.Next() {
+		batch, err := bItr.Value()
+		if err != nil {
+			return err
+		}
+
+		if _, exists := decimalPlaces[batch.Id]; exists {
+			continue
+		}
+
+		cItr, err := ss.ClassInfoTable().List(ormCtx, api.ClassInfoPrimaryKey{})
+		if err != nil {
+			return err
+		}
+
+		for cItr.Next() {
+			class, err := cItr.Value()
+			if err != nil {
+				return err
+			}
+
+			if class.Id == classIds[batch.ProjectId] {
+				decimalPlaces[batch.Id] = abbrevToPrecision[class.CreditType]
+			}
+		}
+
+		cItr.Close()
+	}
+
+	calSupplies := make(map[uint64]math.Dec) // map of batchID to calculated supply
+	supplies := make(map[uint64]math.Dec)    // map of batchID to actual supply
+	bsItr, err := ss.BatchSupplyTable().List(ormCtx, api.BatchSupplyPrimaryKey{})
+	if err != nil {
+		return err
+	}
+	defer bsItr.Close()
+
+	for bsItr.Next() {
+		batchSupply, err := bsItr.Value()
+		if err != nil {
+			return err
+		}
+
+		tSupply := math.NewDecFromInt64(0)
+		rSupply := math.NewDecFromInt64(0)
+		if batchSupply.TradableAmount != "" {
+			tSupply, err = math.NewNonNegativeFixedDecFromString(batchSupply.TradableAmount, decimalPlaces[batchSupply.BatchId])
+			if err != nil {
+				return err
+			}
+		}
+		if batchSupply.RetiredAmount != "" {
+			rSupply, err = math.NewNonNegativeFixedDecFromString(batchSupply.RetiredAmount, decimalPlaces[batchSupply.BatchId])
+			if err != nil {
+				return err
+			}
+		}
+
+		total, err := math.SafeAddBalance(tSupply, rSupply)
+		if err != nil {
+			return err
+		}
+
+		supplies[batchSupply.BatchId] = total
+	}
+
+	// calculate credit batch supply from genesis tradable and retired balances
+	if err := calculateSupply(ormCtx, decimalPlaces, ss, calSupplies); err != nil {
+		return err
+	}
+
+	if err := validateSupply(calSupplies, supplies); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func calculateSupply(ctx context.Context, decimalPlaces map[uint64]uint32, ss api.StateStore, calSupply map[uint64]math.Dec) error {
```

Comment:
similar to above, it would be nice to have the variable form KeyToValue for maps, its not immediately clear to devs what the key is supposed to be in this function

Reply:
It looks like this might have been overlooked in most recent changes.

- `decimalPlaces` should be updated to `batchIdToPrecision`
- `calSupply` should be updated to `batchIdToCalSupply`